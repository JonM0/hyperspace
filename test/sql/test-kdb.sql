CREATE TABLE t (a point4d);

CREATE INDEX t_a_kdbspgist ON t USING spgist(a point4d_kdb);

INSERT INTO t
VALUES ('(0,0,0,1)'),
    ('(0,1,0,0)'),
    ('(0,0,2,2)'),
    ('(1,1,0.5,3)'),
    ('(2,1,2,4)'),
    ('(8,0,2,1)'),
    ('(0,0,2,2)'),
    ('(1,1,0.5,3)'),
    ('(3,-9,2,32)');

SET enable_seqscan = false;

EXPLAIN
SELECT a
FROM t
WHERE a <@ box4d '((0,0,0,0), (1,1,1,1))';

SELECT a
FROM t
WHERE a <@ box4d '((0,0,0,0), (1,1,1,1))';

EXPLAIN
SELECT a
FROM t
WHERE a = '(0,0,0,1)'::point4d;

SELECT a
FROM t
WHERE a = '(0,0,0,1)'::point4d;

EXPLAIN
SELECT a,
    a <->point4d '(0,0,2.34,0)' AS d
FROM t
ORDER BY d;

SELECT a,
    a <->point4d '(0,0,2.34,0)' AS d
FROM t
ORDER BY d;

EXPLAIN
SELECT a,
    a <->point4d '(0,0,0,0)' AS d1,
    a <->point4d '(0,1.12,0,0)' AS d2
FROM t
ORDER BY d1,
    d2;

SELECT a,
    a <->point4d '(0,0,0,0)' AS d1,
    a <->point4d '(0,1.12,0,0)' AS d2
FROM t
ORDER BY d1,
    d2;

EXPLAIN
SELECT a,
    a <->point4d '(0,0,0,0)' AS d1,
    a <->point4d '(0,1.12,0,0)' AS d2
FROM t
WHERE a <@ box4d '((0,0,0,0), (1,1,inf,inf))'
ORDER BY d1,
    d2;

SELECT a,
    a <->point4d '(0,0,0,0)' AS d1,
    a <->point4d '(0,1.12,0,0)' AS d2
FROM t
WHERE a <@ box4d '((0,0,0,0), (1,1,inf,inf))'
ORDER BY d1,
    d2;

SET enable_seqscan = TRUE;
DROP INDEX t_a_kdbspgist;

SELECT a,
    a <->point4d '(0,0,2.34,0)' AS d
FROM t
ORDER BY d;

SELECT a,
    a <->point4d '(0,0,0,0)' AS d1,
    a <->point4d '(0,1.12,0,0)' AS d2
FROM t
ORDER BY d1,
    d2;


DROP TABLE t;