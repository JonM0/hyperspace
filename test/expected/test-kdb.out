CREATE TABLE t (a point4d);
CREATE INDEX t_a_kdbspgist ON t USING spgist(a point4d_kdb);
INSERT INTO t
VALUES ('(0,0,0,1)'),
    ('(0,1,0,0)'),
    ('(0,0,2,2)'),
    ('(1,1,0.5,3)'),
    ('(2,1,2,4)'),
    ('(8,0,2,1)'),
    ('(0,0,2,2)'),
    ('(1,1,0.5,3)'),
    ('(3,-9,2,32)');
SET enable_seqscan = false;
EXPLAIN
SELECT a
FROM t
WHERE a <@ box4d '((0,0,0,0), (1,1,1,1))';
                                  QUERY PLAN                                   
-------------------------------------------------------------------------------
 Bitmap Heap Scan on t  (cost=13.41..31.91 rows=680 width=32)
   Recheck Cond: (a <@ '((0,0,0,0),(1,1,1,1))'::box4d)
   ->  Bitmap Index Scan on t_a_kdbspgist  (cost=0.00..13.24 rows=680 width=0)
         Index Cond: (a <@ '((0,0,0,0),(1,1,1,1))'::box4d)
(4 rows)

SELECT a
FROM t
WHERE a <@ box4d '((0,0,0,0), (1,1,1,1))';
     a     
-----------
 (0,0,0,1)
 (0,1,0,0)
(2 rows)

EXPLAIN
SELECT a
FROM t
WHERE a = '(0,0,0,1)'::point4d;
                                 QUERY PLAN                                 
----------------------------------------------------------------------------
 Bitmap Heap Scan on t  (cost=4.20..14.34 rows=7 width=32)
   Recheck Cond: (a = '(0,0,0,1)'::point4d)
   ->  Bitmap Index Scan on t_a_kdbspgist  (cost=0.00..4.20 rows=7 width=0)
         Index Cond: (a = '(0,0,0,1)'::point4d)
(4 rows)

SELECT a
FROM t
WHERE a = '(0,0,0,1)'::point4d;
     a     
-----------
 (0,0,0,1)
(1 row)

EXPLAIN
SELECT a,
    a <->point4d '(0,0,2.34,0)' AS d
FROM t
ORDER BY d;
                                   QUERY PLAN                                    
---------------------------------------------------------------------------------
 Index Only Scan using t_a_kdbspgist on t  (cost=0.14..79.34 rows=1360 width=40)
   Order By: (a <-> '(0,0,2.34,0)'::point4d)
(2 rows)

SELECT a,
    a <->point4d '(0,0,2.34,0)' AS d
FROM t
ORDER BY d;
      a      |         d          
-------------+--------------------
 (0,0,2,2)   |   2.02869416127715
 (0,0,2,2)   |   2.02869416127715
 (0,0,0,1)   | 2.5447200238926087
 (0,1,0,0)   | 2.5447200238926087
 (1,1,0.5,3) |  3.792835351026986
 (1,1,0.5,3) |  3.792835351026986
 (2,1,2,4)   |  4.595171378740949
 (8,0,2,1)   |  8.069423771249097
 (3,-9,2,32) |  33.37837024181978
(9 rows)

EXPLAIN
SELECT a,
    a <->point4d '(0,0,0,0)' AS d1,
    a <->point4d '(0,1.12,0,0)' AS d2
FROM t
ORDER BY d1,
    d2;
                                   QUERY PLAN                                    
---------------------------------------------------------------------------------
 Index Only Scan using t_a_kdbspgist on t  (cost=0.14..86.14 rows=1360 width=48)
   Order By: ((a <-> '(0,0,0,0)'::point4d) AND (a <-> '(0,1.12,0,0)'::point4d))
(2 rows)

SELECT a,
    a <->point4d '(0,0,0,0)' AS d1,
    a <->point4d '(0,1.12,0,0)' AS d2
FROM t
ORDER BY d1,
    d2;
      a      |         d1         |         d2         
-------------+--------------------+--------------------
 (0,1,0,0)   |                  1 | 0.1200000000000001
 (0,0,0,1)   |                  1 | 1.5014659503298768
 (0,0,2,2)   | 2.8284271247461903 |  3.042104534693047
 (0,0,2,2)   | 2.8284271247461903 |  3.042104534693047
 (1,1,0.5,3) | 3.3541019662496847 | 3.2038102315836374
 (1,1,0.5,3) | 3.3541019662496847 | 3.2038102315836374
 (2,1,2,4)   |                  5 |  4.900448959024061
 (8,0,2,1)   |  8.306623862918075 |  8.381789785004155
 (3,-9,2,32) | 33.436506994600975 |  33.75521293074597
(9 rows)

EXPLAIN
SELECT a,
    a <->point4d '(0,0,0,0)' AS d1,
    a <->point4d '(0,1.12,0,0)' AS d2
FROM t
WHERE a <@ box4d '((0,0,0,0), (1,1,inf,inf))'
ORDER BY d1,
    d2;
                                   QUERY PLAN                                   
--------------------------------------------------------------------------------
 Index Only Scan using t_a_kdbspgist on t  (cost=0.14..66.84 rows=680 width=48)
   Index Cond: (a <@ '((0,0,0,0),(1,1,Infinity,Infinity))'::box4d)
   Order By: ((a <-> '(0,0,0,0)'::point4d) AND (a <-> '(0,1.12,0,0)'::point4d))
(3 rows)

SELECT a,
    a <->point4d '(0,0,0,0)' AS d1,
    a <->point4d '(0,1.12,0,0)' AS d2
FROM t
WHERE a <@ box4d '((0,0,0,0), (1,1,inf,inf))'
ORDER BY d1,
    d2;
      a      |         d1         |         d2         
-------------+--------------------+--------------------
 (0,1,0,0)   |                  1 | 0.1200000000000001
 (0,0,0,1)   |                  1 | 1.5014659503298768
 (0,0,2,2)   | 2.8284271247461903 |  3.042104534693047
 (0,0,2,2)   | 2.8284271247461903 |  3.042104534693047
 (1,1,0.5,3) | 3.3541019662496847 | 3.2038102315836374
 (1,1,0.5,3) | 3.3541019662496847 | 3.2038102315836374
(6 rows)

explain
SELECT a,
    a <->point4d '(0,0,2.34,0)' AS d
FROM t
where a <@ circle4d '((0,0,2.34,0), 2.1)'
ORDER BY d;
                                   QUERY PLAN                                   
--------------------------------------------------------------------------------
 Index Only Scan using t_a_kdbspgist on t  (cost=0.14..63.45 rows=680 width=40)
   Index Cond: (a <@ '((0,0,2.34,0),2.1)'::circle4d)
   Order By: (a <-> '(0,0,2.34,0)'::point4d)
(3 rows)

SELECT a,
    a <->point4d '(0,0,2.34,0)' AS d
FROM t
where a <@ circle4d '((0,0,2.34,0), 2.1)'
ORDER BY d;
     a     |        d         
-----------+------------------
 (0,0,2,2) | 2.02869416127715
 (0,0,2,2) | 2.02869416127715
(2 rows)

SELECT a,
    a <->point4d '(1,1,1,1)'
FROM t
where a <@ circle4d '((1,1,1,1), 4.3)' and a <@ box4d '((0,0,0,0), (3,3,2,2))';
     a     |      ?column?      
-----------+--------------------
 (0,0,0,1) | 1.7320508075688772
 (0,1,0,0) | 1.7320508075688772
 (0,0,2,2) |                  2
 (0,0,2,2) |                  2
(4 rows)

SET enable_seqscan = TRUE;
DROP INDEX t_a_kdbspgist;
SELECT a,
    a <->point4d '(0,0,2.34,0)' AS d
FROM t
ORDER BY d;
      a      |         d          
-------------+--------------------
 (0,0,2,2)   |   2.02869416127715
 (0,0,2,2)   |   2.02869416127715
 (0,1,0,0)   | 2.5447200238926087
 (0,0,0,1)   | 2.5447200238926087
 (1,1,0.5,3) |  3.792835351026986
 (1,1,0.5,3) |  3.792835351026986
 (2,1,2,4)   |  4.595171378740949
 (8,0,2,1)   |  8.069423771249097
 (3,-9,2,32) |  33.37837024181978
(9 rows)

SELECT a,
    a <->point4d '(0,0,0,0)' AS d1,
    a <->point4d '(0,1.12,0,0)' AS d2
FROM t
ORDER BY d1,
    d2;
      a      |         d1         |         d2         
-------------+--------------------+--------------------
 (0,1,0,0)   |                  1 | 0.1200000000000001
 (0,0,0,1)   |                  1 | 1.5014659503298768
 (0,0,2,2)   | 2.8284271247461903 |  3.042104534693047
 (0,0,2,2)   | 2.8284271247461903 |  3.042104534693047
 (1,1,0.5,3) | 3.3541019662496847 | 3.2038102315836374
 (1,1,0.5,3) | 3.3541019662496847 | 3.2038102315836374
 (2,1,2,4)   |                  5 |  4.900448959024061
 (8,0,2,1)   |  8.306623862918075 |  8.381789785004155
 (3,-9,2,32) | 33.436506994600975 |  33.75521293074597
(9 rows)

DROP TABLE t;
